gitpod-io/gitpod
# Add the Browser Terminal as an IDE

## Description
Adds Xterm.js as a web alternative to connecting to your workspace via SSH from your terminal. The code can be found at https://github.com/gitpod-io/xterm-web-ide and the Docker image is at [ide/xterm-web](https://console.cloud.google.com/gcr/images/gitpod-core-dev/eu/build/ide/xterm-web).

> **Note**: The original PR is at https://github.com/gitpod-io/gitpod/pull/8464, but it was merged incorrectly. 

> **Note**: this relies on https://github.com/gitpod-io/gitpod/pull/17240 being merged first

Preview environment 🌐: https://ide-browser-terminal.preview.gitpod-dev.com/workspaces

## How to test
1. Open the preview environment
2. In the preferences, notice there is no Xterm IDE
3. In [Configcat](https://app.configcat.com/), include yourself in the enable-rule for `experimentalIdes` (non-prod)
4. Wait a couple of minutes (🫖 time)
5. Reload and see the Terminal IDE magically appear
6. Create a workspace and have a blast

## Useful commands

Re-pull the IDE: 
```sh
kubectl rollout restart deployment/ide-service
```

## Current limitations

- [x] Does not work on mobile [fixed]
	<details><summary>iPad demo</summary>
		
	https://user-images.githubusercontent.com/29888641/227277864-891126e7-de8a-4038-a801-9ea669167952.mp4
		

	</details> 

## Release Notes

```release-note
NONE
```
⬆️ we can mention it after we finish the internal dog food

## Documentation


## Todo

- [x] `gp stop` closes xterm.js editor and shows the "Workspace Stopping" screen 
- [x] `gp open` works
- [x] `gp preview` works
- [x] Ensure heartbreating works
- [x] Move Docker image from DockerHub to our GCP registry

### Nice-to-haves
- [ ] Local echo [[source](https://github.com/gitpod-io/openvscode-server/blob/gp-code/main/src/vs/workbench/contrib/terminal/browser/xterm/terminalTypeAheadAddon.ts)]

## Build Options:

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

</details>

<details>
<summary>Installer Options</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
